micro_http/src/server.rs

```diff
@@ -0,0 +1,752 @@
+use std::collections::HashMap;
+
+static SERVER_FULL_ERROR_MESSAGE: &'static [u8] = b"HTTP/1.1 503\r\n\
+                                            Server: Firecracker API\r\n\
+                                            Connection: close\r\n\
+                                            Content-Length: 14\r\n\r\nServer is full.";
+const MAX_CONNECTIONS: usize = 10;
+
+type Result<T> = std::result::Result<T, ServerError>;
+
+/// Wrapper over `Request` which adds an identification token.
+pub struct ServerRequest {
+    /// Inner request.
+    pub request: Request,
+    /// Identification token.
+    id: u64,
+}
+
+impl ServerRequest {
+    /// Creates a new `RequestWithID` object from an existing `Request`,
+    /// adding an identification token.
+    pub fn new(request: Request, id: u64) -> Self {
+        ServerRequest { request, id }
+    }
+
+    /// Returns a reference to the inner request.
+    pub fn inner(&self) -> &Request {
+        &self.request
+    }
+
+    /// Docs needed.
+    pub fn process<F>(&self, callable: F) -> ServerResponse
+    where
+        F: Fn(&Request) -> Response,
+    {
+        let http_response = callable(self.inner());
+        ServerResponse::new(http_response, self.id)
+    }
+}
+
+/// Wrapper over `Response` which adds an identification token.
+pub struct ServerResponse {
+    /// Inner response.
+    response: Response,
+    /// Identification token.
+    id: u64,
+}
+
+impl ServerResponse {
+    fn new(response: Response, id: u64) -> ServerResponse {
+        ServerResponse { response, id }
+    }
+}
+
+/// Describes the state of the connection as far as data exchange
+/// on the stream is concerned.
+#[derive(PartialOrd, PartialEq)]
+enum ClientConnectionState {
+    AwaitingIncoming,
+    AwaitingOutgoing,
+    Closed,
+}
+
+/// Wrapper over `HttpConnection` which keeps track of yielded
+/// requests and absorbed responses.
+struct ClientConnection<T> {
+    /// The `HttpConnection` object which handles data exchange.
+    connection: HttpConnection<T>,
+    /// The state of the connection in the `epoll` structure.
+    state: ClientConnectionState,
+    /// Represents the difference between yielded requests and
+    /// absorbed responses.
+    /// This has to be `0` if we want to drop the connection.
+    in_flight_response_count: u32,
+}
+
+impl<T: Read + Write> ClientConnection<T> {
+    fn new(connection: HttpConnection<T>) -> Self {
+        ClientConnection {
+            connection,
+            state: ClientConnectionState::AwaitingIncoming,
+            in_flight_response_count: 0,
+        }
+    }
+
+    fn read(&mut self) -> Result<Vec<Request>> {
+        // Data came into the connection.
+        let mut parsed_requests = vec![];
+        match self.connection.try_read() {
+            Err(ConnectionError::ConnectionClosed) => {
+                // Connection timeout.
+                self.state = ClientConnectionState::Closed;
+                return Ok(parsed_requests);
```

Comment:
`HttpConnection` propagates the `ConnectionClosed` error upwards, as intended. However, `HttpServer` doesn't because it is handled internally. If the user closed the connection the server will then quietly drop the connection. When the server handles notifications from a connection it will propagate any errors from that connection, but `read` will never return a `ClosedConnection` error, as it is handled separately.

Reply:
The high-level behavior question @andreeaflorescu is posing here is:
If the server receives `N` correct requests and then the client closes its connection (all of these in a short burst), should we:
1. Just drop those requests; don't let the user of the server even know of them (since the client has closed its end, there's no way to provide a response anyway),
or
2. Propagate/deliver those requests and allow the user of the server to process them and provide responses back to the server, which the server silently drops since there is no longer a client on the other side. (This is the currently implemented behaviour)

Reply:
There only ways to know if a client has closed the connection is to either read everything from the stream and then reach EOF, or try writing something to the stream and fail. The former is what we are doing now, while the latter is impossible for this type of connection because we would send invalid bytes as a response.

Reply:
I have replaced the return value on that matching arm with an empty vector. Furthermore, I have added a commit in which we now drop unyielded requests when the client closes the connection.